wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,180 @@
+//! A lint rule to ensure each output is documented in `meta`.
+
+use wdl_ast::v1::MetadataSection;
+use wdl_ast::v1::OutputSection;
+use wdl_ast::v1::TaskDefinition;
+use wdl_ast::v1::TaskOrWorkflow;
+use wdl_ast::v1::WorkflowDefinition;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::ToSpan;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the non-matching output rule.
+const ID: &str = "NonmatchingOutput";
+
+/// Creates a "non-matching output" diagnostic.
+fn nonmatching_output(span: Span, name: String, context: TaskOrWorkflow) -> Diagnostic {
+    let (ty, item_name) = match context {
+        TaskOrWorkflow::Task(t) => ("task", t.name().as_str().to_string()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name().as_str().to_string()),
+    };
+
+    Diagnostic::warning(format!(
+        "output `{name}` is missing from `meta` section in {ty} `{item_name}`"
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(format!(
+        "add output (`{name}`) key to `outputs` documentation in `meta` section"
+    ))
+}
+
+/// Creates a missing outputs in meta diagnostic.
+fn missing_outputs_in_meta(span: Span, context: TaskOrWorkflow) -> Diagnostic {
+    let (ty, name) = match context {
+        TaskOrWorkflow::Task(t) => ("task", t.name().as_str().to_string()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name().as_str().to_string()),
+    };
+
+    Diagnostic::warning(format!(
+        "`outputs` key missing in `meta` section in {ty} `{name}`"
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix("add `outputs` key to `meta` section")
+}
+
+/// Detects non-matching outputs.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct NonmatchingOutputRule;
+
+impl Rule for NonmatchingOutputRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that each output field is documented in the meta section."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "The meta section should have an output key and keys with descriptions for each output of \
+         the task/workflow. These must match exactly. i.e. for each named output of a task or \
+         workflow, there should be an entry under meta.output with that same name. Additionally, \
+         these entries should be in the same order (that order is up to the developer to decide). \
+         No extraneous output entries are allowed. There should not be any blank lines inside the \
+         entire meta section."
```

Comment:
```suggestion
         No extraneous output entries are allowed."
```
As this sentence isn't being enforced by this rule, it shouldn't be mentioned here. (Although maybe once that lint rule gets written it could be "linked" (per #116 )).

Reply:
I typically just copy/paste the description from #11. So this isn't accurate at all. I'm not enforcing `No extraneous output entries are allowed` either. 

The spacing statement seems like it should be in a different rule anyway.

Reply:
I'm also not enforcing that the orders are the same.

Reply:
IMO at least the "no extraneous entries" check belongs within this rule. Probably the order as well. We've discussed (without deciding anything) that maybe `input` and `parameter_meta` sections could be in different orders. There's no rule in #11 about ordering the output section. Seeing as how that's completely "customizable" I think enforcing the same order in the `meta.output` makes sense.

Reply:
I ended up checking both lists and reporting extraneous entries in either list. Then if no extra/missing entries are found, I compare the order. It's a little cumbersome because if something is missing and the `outputs` key is out-of-order, you won't get the out-of-order message until you fix the first one. I toyed with trying to do something to remove the extra/missing entries and check the resulting ordering, but I think that's going to be cumbersome and also potentially confusing to the users (and possibly error-prone).